src/wallet/rpcwallet.cpp

```diff
@@ -1160,26 +1161,55 @@ static UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bo
         if (nDepth < nMinDepth)
             continue;
 
-        for (const CTxOut& txout : wtx.tx->vout)
+        if (received)
         {
-            CTxDestination address;
-            if (!ExtractDestination(txout.scriptPubKey, address))
-                continue;
+            for (const CTxOut& txout : wtx.tx->vout)
+            {
+                CTxDestination address;
+                if (!ExtractDestination(txout.scriptPubKey, address))
+                    continue;
+
+                if (has_filtered_address && !(filtered_address == address)) {
+                    continue;
+                }
+
+                isminefilter mine = IsMine(*pwallet, address);
+                if(!(mine & filter))
+                    continue;
 
-            if (has_filtered_address && !(filtered_address == address)) {
-                continue;
+                tallyitem& item = mapTally[address];
+                item.nAmount += txout.nValue;
+                item.nConf = std::min(item.nConf, nDepth);
+                item.txids.push_back(wtx.GetHash());
+                if (mine & ISMINE_WATCH_ONLY)
+                    item.fIsWatchonly = true;
             }
+        }
+        else
+        {
+            for (const CTxIn& txin : wtx.tx->vin)
+            {
+                const COutPoint& prevout = txin.prevout;
+                const CTxOut& txout = pwallet->GetWalletTx(prevout.hash)->tx->vout[prevout.n];
```

Comment:
Are you sure about this?

Reply:
I assumed there's no need for tests since this got a NACK.

I was going to check it with tests, but if I undersand it correctly, since we're interested only in wallet transactions, checking if result of `GetWalletTx` is non-`nullptr`would be enough? 
```suggestion
                const CWalletTx* iwtx = pwallet->GetWalletTx(prevout.hash);
                if (!iwtx) continue;
                const CTxOut& txout = iwtx->tx->vout[prevout.n];
```
Or am I missing something again?